src/main/java/com/google/sps/servlets/SavedRecipeServlet.java

```diff
@@ -0,0 +1,131 @@
+
+import java.util.Arrays;
+import java.util.ArrayList;
+
+/** Servlet that posts and gets saved recipe information in Datastore */
+@WebServlet("/saved-recipe")
+public class SavedRecipeServlet extends HttpServlet {
+  
+  /** Gets SavedRecipe using the recipe's id */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    JSONObject responseMap = new JSONObject();
+
+    // get recipeId from request parameter
+    Long recipeId = Long.parseLong(request.getParameter("recipeId"));
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("SavedRecipe")
+      .setFilter(new Query.FilterPredicate("recipeId", Query.FilterOperator.EQUAL, recipeId));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      responseMap.put("recipeIsSaved", false);
+    } else {
+      String title = (String) entity.getProperty("title");
+      String imageUrl = (String) entity.getProperty("imageUrl");
+      String sourceUrl = (String) entity.getProperty("sourceUrl");
+      Long servings = (Long) entity.getProperty("servings");
+      Long readyInMinutes = (Long) entity.getProperty("readyInMinutes");
+      ArrayList<String> dietaryNeedsStrings = (ArrayList<String>) entity.getProperty("dietaryNeeds");
+      
+      // convert string to Diet enum because datastore stores dietaryNeeds as a list of strings
+      ArrayList<SavedRecipe.Diet> dietaryNeeds = new ArrayList<>();
+      for (String dietString: dietaryNeedsStrings) {
+        switch(dietString) {
+          case "VEGETARIAN":
+            dietaryNeeds.add(SavedRecipe.Diet.VEGETARIAN);
+            break;
+          case "VEGAN":
+            dietaryNeeds.add(SavedRecipe.Diet.VEGAN);
+            break;
+          case "GLUTENFREE":
+            dietaryNeeds.add(SavedRecipe.Diet.GLUTENFREE);
+            break;
+          case "DAIRYFREE":
+            dietaryNeeds.add(SavedRecipe.Diet.DAIRYFREE);
+            break;
+          default: break;
+        }
+      }
+      SavedRecipe savedRecipeObject = SavedRecipe.builder().setId(recipeId).setTitle(title).
+        setImage(imageUrl).setSourceUrl(sourceUrl).setServings(servings).setReadyInMinutes(readyInMinutes).setDietaryNeeds(dietaryNeeds).build();
+          
+      responseMap.put("savedRecipe", savedRecipeObject);
+      responseMap.put("recipeIsSaved", true);
+    }
+    
+    Gson gson = new Gson();   
+    String json = gson.toJson(responseMap);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /** saves recipe information in datastore */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
```

Comment:
Could you elaborate what is the use case for saving recipe?

I am a bit confused, is this recipe saved and attached to the specific user or is it shared so that any user could save a recipe which is shared and readable among all users? If the latter, since all the recipe info is read from frontend request instead of reading from the Spoonacular API, does it mean any user could hack the request and change the info stored for the specific recipe-id?

Reply:
By saving the recipe information, it provides a way to obtain the information needed to display the recipes when searching by tags without taking up more query points for spoonacular. Since the recipe information doesn't change from user to user I thought that the information could be shared among users, but I can understand your concern about how the users could hack the request.